### PR TITLE
VizLegend: Represent line style in series legend and tooltip

### PR DESCRIPTION
**Pull request by domasx2**
**What is this feature?**

This PR makes panel legend & viz tooltip  color indicator also represent line style of a series: solid, dotted or dashed. Only works if gradient is not set, because we can't represent it in dashed/ dotted form with current appraoch.

dot-dash-cmp
dot-dash-cmp

**Why do we need this feature?**

This is especially useful with time comparison feature where UX calls for comparison series to have same color but different line style. Eg, app o11y uses dashed lines for previous period series. Without legend representing line style it is harder to distinguish legend items.

**Who is this feature for?**

Dashboard users, plugins.

**Special notes for your reviewer:**

I went ahead and made viz tooltip reuse `SeriesIcon` component for color indicator

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by leeoniya** [comment]
would it be possible to avoid extra DOM here? especially for tooltip that has to re-render up to 60 times per sec, this will add a lot of overhead for something that we might be able to do with a single SVG element.

**Comment by domasx2** [comment]
> would it be possible to avoid extra DOM here? especially for tooltip that has to re-render up to 60 times per sec, this will add a lot of overhead for something that we might be able to do with a single React.memo'd SVG element or clever CSS.

I'll try. What's the case where dom is recreated 60 times per second though? :-O would React.memo'ing the divs work too?

**Comment by leeoniya** [comment]
> I'll try. What's the case where dom is recreated 60 times per second though?

sorry, i usually mean react's v-tree, which very unfree to re-create on ever mousemove. imagine you have 30 series in a tooltip. this adds up very fast.

**Comment by domasx2** [comment]
I removed extra divs and refactored with border styling, which gives almost the same result in Chrome 
image

Will test with more browsers tomorrow

File: packages/grafana-ui/src/components/VizLegend/SeriesIcon.tsx [hunk 1]
```diff
@@ -29,28 +31,47 @@ export const SeriesIcon = React.memo(
...
 
-const getStyles = () => ({
+const getStyles = (theme: GrafanaTheme2) => ({
+  container: css({
+    marginRight: '8px',
+    display: 'inline-block',
+    width: '14px',
+  }),
+  dot: css({
+    borderTop: 'dotted 3.99px',
```
Comment (leeoniya) [inline packages/grafana-ui/src/components/VizLegend/SeriesIcon.tsx:65]: 3.99, huh? :sweat_smile:

**Comment by leeoniya** [comment]
> I removed extra divs and refactored with border styling, which gives almost the same result in Chrome

thanks :)

File: packages/grafana-ui/src/components/VizLegend/SeriesIcon.tsx [hunk 2]
```diff
@@ -29,28 +31,47 @@ export const SeriesIcon = React.memo(
...
+      lineStyle?.fill === 'dot' ? styles.dot : lineStyle?.fill === 'dash' ? styles.dash : styles.solid;
+
+    const colorStyle: CSSProperties =
+      lineStyle?.fill === 'dot' || lineStyle?.fill === 'dash'
+        ? {
+            borderTopColor: cssColor,
+          }
+        : {
+            backgroundColor: cssColor,
+          };
```
Comment (leeoniya) [inline packages/grafana-ui/src/components/VizLegend/SeriesIcon.tsx:44]: we can probably use a fat border for all cases (even solid)?

File: packages/grafana-ui/src/components/VizLegend/SeriesIcon.tsx (hunk 1 above)
Comment (domasx2) [reply packages/grafana-ui/src/components/VizLegend/SeriesIcon.tsx:65]: it feels a lot more affordable than 4 :D 

using `4` makes only 2 dots show up with a large gap *on my machine*, hence the odd "hack"

**Comment by domasx2** [comment]
Problem with border based solution is that it's not consistent across render engines. In firefox: 
image

File: packages/grafana-ui/src/components/VizLegend/SeriesIcon.tsx (hunk 2 above)
Comment (domasx2) [reply packages/grafana-ui/src/components/VizLegend/SeriesIcon.tsx]: done!

**Comment by domasx2** [comment]
So border did not work out, too inconsistent across borwsers.
Refactored to third attempt which seems to work well:  repeated background images using `radial-gradient` and `linear-gradient`. This gives consistent result across all browsers.
Caveat is that it doesn't render gradient too, so in case gradient is set it will default to previous looks

**Review by leeoniya — APPROVED** [review]
awesome! :heart: :shipit:

**Pull request merged**